nvflare/fuel/f3/sfm/conn_manager.py

```diff
@@ -0,0 +1,451 @@
+        """
+
+        if endpoint.name == self.local_endpoint.name:
+            self.send_loopback_message(endpoint, app_id, headers, payload)
+            return
+
+        sfm_endpoint = self.sfm_endpoints.get(endpoint.name)
+        if not sfm_endpoint:
+            raise CommError(CommError.CLOSED, f"Endpoint {endpoint.name} not available, may be disconnected")
+
+        state = sfm_endpoint.endpoint.state
+        if state != EndpointState.READY:
+            raise CommError(CommError.NOT_READY, f"Endpoint {endpoint.name} is not ready: {state}")
+
+        stream_id = sfm_endpoint.next_stream_id()
+
+        # When multiple connections, round-robin by stream ID
+        sfm_conn = sfm_endpoint.get_connection(stream_id)
+        if not sfm_conn:
+            log.error("Logic error, ready endpoint has no connections")
+            raise CommError(CommError.ERROR, f"Endpoint {endpoint.name} has no connection")
+
+        # TODO: If multiple connections, should retry a diff connection on errors
+        start = time.perf_counter()
+
+        sfm_conn.send_data(app_id, stream_id, headers, payload)
+
+        self.send_frame_stats.record_value(
+            category=sfm_conn.conn.connector.driver.get_name(), value=time.perf_counter() - start
+        )
+
+    def register_message_receiver(self, app_id: int, receiver: MessageReceiver):
+        if self.receivers.get(app_id):
+            raise CommError(CommError.BAD_DATA, f"Receiver for app {app_id} is already registered")
+
+        self.receivers[app_id] = receiver
+
+    def add_endpoint_monitor(self, monitor: EndpointMonitor):
+        self.monitors.append(monitor)
+
+    # Internal methods
+
+    def start_connector(self, connector: ConnectorInfo):
+        """Start connector in a new thread"""
+
+        if connector.started:
+            return
+
+        log.info(f"Connector {connector} is starting")
+
+        self.conn_mgr_executor.submit(self.start_connector_task, connector)
+
+    @staticmethod
+    def start_connector_task(connector: ConnectorInfo):
+        """Start connector in a new thread
+        This function will loop as long as connector is not stopped
+        """
+
+        connector.started = True
+        if connector.mode == Mode.ACTIVE:
+            starter = connector.driver.connect
+        else:
+            starter = connector.driver.listen
+
+        wait = INIT_WAIT
+        while not connector.stopping:
+            start_time = timeit.timeit()
+            try:
+                starter(connector)
+            except Exception as ex:
+                log.error(f"Connector {connector} failed with exception {type(ex).__name__}: {ex}")
+                log.debug(traceback.format_exc())
+
+            if connector.stopping:
+                log.debug(f"Connector {connector} has stopped")
+                break
+
+            # After a long run, resetting wait
+            run_time = timeit.timeit() - start_time
```

Comment:
I think this is not the right way to use timeit. timeit.timeit() is to wrap the execution codes as a parameter for the timeit to measure the execution time. It may gives a little more accurate time. But we can not simply use timeit.timeit() to replace time.time(). For your execution logic, it still needs to use time.time().

BTW, in most of our use cases, we are not measuring the execution time in milliseconds level,  time.time() usage is simply enough.

Reply:
You are correct. That's the wrong way to use it. Reverted. No need for accuracy in this use-case.